### ISSUE TITLE
Render the cracked node counterpart instead of the rebuilding the whole mapblock mesh

### ISSUE THREAD

**Issue by Andrey2470T** [minetest/minetest#15287]
### Problem

Currently when the crack animation is playing (the crack level is different from -1) on the node which a player is punching, Minetest reconstructs the whole mapblock mesh in which the node locates and also adjacent ones to that locating at the corners. The reconstructing the mapblocks meshes happen twice (at the beggining of the animation and at the end). That makes some delays in the animation updates that can be pretty tangible especially if the mapblocks meshes have a much geometry with the complex meshes each of which contains hundreds of vertices to render afterwards. The following example shows the mapblock fully consisting of the high-poly `homedecor:bathroom_set` mesh nodes and how the punches on them from the player affects the framerate and the lag in the crack animations:

https://github.com/user-attachments/assets/73c827bc-435c-436f-8d9f-43f96525254a

It's seen that the crack frames get updated not immediately, but only after ~1 s. Except that FPS drops 15-20 below at the punches time.

I've done some profiling also showing how much time (measured in us) goes on the MapBlockMesh constructor call for the case above:
```
// Beginning of the crack animation
INFO[MeshUpdate]: Rebuild mapblock mesh took 299363us
// End of the crack animation
INFO[MeshUpdate]: Rebuild mapblock mesh took 303053us
```

That is pretty much time, about 300 ms. In the simple mapblock meshes consisting of the regular nodes like a grass, stone in MTG the time for the rebuilding goes somewhere in 150 times less, some example:
```
// Beginning of the crack animation
INFO[MeshUpdate]: Rebuild mapblock mesh took 2352us
// End of the crack animation
INFO[MeshUpdate]: Rebuild mapblock mesh took 641us
```

Also, it is worth to mention that the crack textures get overlaid on the tile layers textures making generate separate textures and cache them in the texture source cache what cause needlessly spending the GPU memory.

### Solutions

Render a single transparent node counterpart with only crack textures atop the cracked node at the same position instead of the reconstructing the mapblock mesh. It will be rendered in the "overlay" OpenGL blend mode:
`glBlendFunc(GL_DST_COLOR, GL_SRC_COLOR)`.
That actually will just add the source and destination colors together. From this approach one can obtain the next benefits:

- Render only the single node instead of the 16x16x16 nodes + 3 adjacent mapblock meshes.
- No need to generate and cache the textures combined from the cracked node textures and cracks ones.
- The blend mode will mix the the underlying and crack texture instead of just overlaying on it such way providing visually a better result:

![image-2](https://github.com/user-attachments/assets/9d16c1cd-1ba5-4a6d-b0c1-24af2d3dd14c)

A few benefits for my atlas PR #15061 :
- No a need to double the atlas texture space for the crack tiles (currently my approach does it).
- Will fix the bug with the crack textures can currently inevitably only overlay on the atlas (obviously it leads to the visible texture distortion).
- No a need to offset the tiles UVs to the crack area inside the atlas.

### Alternatives

I don't have an idea for other alternative solutions.

**Comment by Desour** [comment]
Can probably also fix:
* #14270
* #5317
* Crack scaled down for smaller than crack (e.g. 1x1) textures. (Haven't found issue.)
* Weird looking crack for nodes with non-trivial uv mapping.

---

Meshgen time heavily depends on the block mesh chunk size setting.
I've seen issues in debug builds.

---

Idk what the right way is to overlay the crack animation (e.g. blending or not blending, or adding as you suggested). But changing behaviour might be controversial.

---

To get the right draw order, I suggest you could insert the crack mesh's triangles into the transparency-sorting bsp tree, and draw it together with the semi-transparent nodes.

---

How do you want to generate the mesh for the crack? A simple cube won't always work.

---

What one could also try is storing node position information together with the vertices, and only remesh the single node (and maybe its neighbors).

**Comment by Andrey2470T** [comment]
> Idk what the right way is to overlay the crack animation (e.g. blending or not blending, or adding as you suggested). But changing behaviour might be controversial.
> 
That could be adjusted by the global setting I assume. In the case if the overlay is denabled there, there won't the blending happened and the transparent mesh with cracks will be drawn just atop the cracked node on the screen. But the problem is those pixels not belonging to the cracks will also just be overlaid and as they are black, the whole cracked mesh will be just black.

> How do you want to generate the mesh for the crack? A simple cube won't always work.
> 
Duplicate the cracked node mesh, set it at the same nodepos where the cracked one locates and shift a bit its position towards to the player pos to prevent the z-fighting. The cracked mesh will be mainly transparent and have only the cracks textures, see how this mesh would look in the game:
![maxresdefault](https://github.com/user-attachments/assets/0adfeb52-e266-404c-bb32-58a87bbeb1dd)

And that duplicate mesh will be rendered in the overlay mode, such way allowing to see-through the actual cracked mesh when it is applied.

> To get the right draw order, I suggest you could insert the crack mesh's triangles into the transparency-sorting bsp tree, and draw it together with the semi-transparent nodes.
>
There is no a need in sorting for the single node duplicate, moreover this mesh would be temporary, existing only during the crack animation. Its mesh can be kept somewhere separately, like the selection mesh. And that should be also rendered separately, probably either after the rendering all mapblocks at the end of the `ClientMap::renderMap` or even in a separate render step.
> What one could also try is storing node position information together with the vertices, and only remesh the single node (and maybe its neighbors).
It is impossible to remesh the single node since its mesh buffers are merged together with other ones by the identical material within the mapblock mesh.

**Comment by Desour** [comment]
> > To get the right draw order, I suggest you could insert the crack mesh's triangles into the transparency-sorting bsp tree, and draw it together with the semi-transparent nodes.
> 
> There is no a need in sorting for the single node duplicate, moreover this mesh would be temporary, existing only during the crack animation. Its mesh can be kept somewhere separately, like the selection mesh. And that should be also rendered separately,probably either after the rendering all mapblocks at the end of the `ClientMap::renderMap` or even in a separate render step.

You do need to sort. Otherwise, you either won't be able to see cracks behind semitransparent surfaces (e.g. behind water) (selection has this issue #5780), or (if you draw the crack before the transparent nodes) you might see weird artifacts when digging semitransparent nodes.

> > What one could also try is storing node position information together with the vertices, and only remesh the single node (and maybe its neighbors).
> 
> It is impossible to remesh the single node since its mesh buffers are merged together with other ones by the identical material within the mapblock mesh.

It would be possible if one stores the node position, as said.

---

Another thing that got into my mind: Cracks currently are not drawn over full transparent pixels. (There really should be a tiledef field for this.)  To get this behaviour, you could sample the texture of the underlying node, and discard appropriately.

**Comment by Andrey2470T** [comment]
>You do need to sort. Otherwise, you either won't be able to see cracks behind semitransparent surfaces (e.g. behind water) (selection has this issue https://github.com/minetest/minetest/issues/5780), or (if you draw the crack before the transparent nodes) you might see weird artifacts when digging semitransparent nodes.
>
Ah, I got it. I assume I will need to save a separate crack mesh inside the corresponding mapblock mesh and then insert its triangles to the general array, thus triggering the bsp tree rebuilding... and still rebuilding when the crack animation has finished. But in any case it won't rebuild the whole mapblock mesh again, so it should be though faster than the current approach.

**Comment by Andrey2470T** [comment]
>It would be possible if one stores the node position, as said.
>
Could you elaborate how you suggest to implement it? How would the node position help to extract the correct mesh buffers from the general mesh dump with their materials and replace the textures in their materials with the cracked variants?

>That could be adjusted by the global setting I assume. In the case if the overlay is denabled there, there won't the blending happened and the transparent mesh with cracks will be drawn just atop the cracked node on the screen. But the problem is those pixels not belonging to the cracks will also just be overlaid and as they are black, the whole cracked mesh will be just black.
>
I mistook, the blending could quite happen, just in another mode (assuming the standard cracks texture has the RGBA format): `glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)`. So that's not the problem.

>Another thing that got into my mind: Cracks currently are not drawn over full transparent pixels. (There really should be a tiledef field for this.) To get this behaviour, you could sample the texture of the underlying node, and discard appropriately.
>
I suppose it won't be the problem in my approach also in both blending modes. Probably that fault is of the incorrect pixel blitting when two textures are overlaid on each other with using the overlay texture modifier and that shouldn't happen with the OpenGL blending.

**Comment by Desour** [comment]
> > It would be possible if one stores the node position, as said.
> 
> Could you elaborate how you suggest to implement it? How would the node position help to extract the correct mesh buffers from the general mesh dump with their materials and replace the textures in their materials with the cracked variants?

You'd store for each mesh buffer's vertex array, from which node pos each vertex came from (e.g. in a separate array, with runlength encoding). Then instead of remeshing the whole chunk, you can copy every vertex from the old meshbuf where the vertex's associated node pos is not the cracked node, and remesh the cracked node.

This would also be useful to make remeshing cheaper when only few nodes in a mapblock change (e.g. when someone opens a door).

(Making a separate, temporary mesh for the crack, as suggested by you, probably is easier though.)

**Comment by Andrey2470T** [comment]
> > > It would be possible if one stores the node position, as said.
> > 
> > 
> > Could you elaborate how you suggest to implement it? How would the node position help to extract the correct mesh buffers from the general mesh dump with their materials and replace the textures in their materials with the cracked variants?
> 
> You'd store for each mesh buffer's vertex array, from which node pos each vertex came from (e.g. in a separate array, with runlength encoding). Then instead of remeshing the whole chunk, you can copy every vertex from the old meshbuf where the vertex's associated node pos is not the cracked node, and remesh the cracked node.
> 
> This would also be useful to make remeshing cheaper when only few nodes in a mapblock change (e.g. when someone opens a door).
> 
> (Making a separate, temporary mesh for the crack, as suggested by you, probably is easier though.)

No, the remeshing won't work correctly if necessary to replace the textures of the cracked node to the cracks ones. Because the meshgen implements the batching by a material and that would replace to the cracks textures in *all* nodes tiles which having the same textures which the cracked node has.

**Comment by Desour** [comment]
> No, the remeshing won't work correctly if necessary to replace the textures of the cracked node to the cracks ones. Because the meshgen implements the batching by a material and that would replace to the cracks textures in _all_ nodes tiles having the same textures which the cracked node has.

The meshgen uses a different material for the crack. (And if it didn't, the issue you're describing would already happen in master. :P)

**Comment by Andrey2470T** [comment]
> > No, the remeshing won't work correctly if necessary to replace the textures of the cracked node to the cracks ones. Because the meshgen implements the batching by a material and that would replace to the cracks textures in _all_ nodes tiles having the same textures which the cracked node has.
> 
> The meshgen uses a different material for the crack. (And if it didn't, the issue you're describing would already happen in master. :P)

That is, does your idea assume the creating a temporary crack material also (you keep the array of the node positions within some mapblock with each from which the certain vertex is associated, e g. through indices, then looping through them finding for the correponding position and then moving the vertices from the original buffers in some new one with that crack material attached, then after the crack animation the vertices get back in their original buffers?). If so, I think such way would fit better for when some node is added/removed or param2 is changed since in such cases it is unnecessary to create the temporary materials, but only add/remove vertices or change their positions. Note the looping through the whole mesh where the vertex count can be millions may be expensive also.

**Comment by nininik0** [comment]
Guys what about the crack texture being weird when the texture is smaller then the crack texture???????????!

**Comment by appgurueu** [comment]
Yes, this can definitely be optimized, but I wonder how important that is. The example in your post is deliberately contrived (and would arguably probably benefit most from some kind of instancing rather than batching). How bad is this phenomenon really for "normal" mapblocks?

Under the assumption that mapblocks have similar average complexities, this single mapblock doesn't matter terribly much in the bigger picture when it comes to throughput. If we can't afford to regenerate a single mapblock mesh, how could we hope to generate the cubically many mapblock meshes around the player in reasonable time? If anything, I would assume this to primarily be a latency issue, wherein regenerating a mapblock might take longer than a frame.

> What one could also try is storing node position information together with the vertices [...]

might very well be a pessimization: You'd be slowing down meshgen for all mapblocks by needing to populate this index just so you can update a single one faster.

**Comment by sfan5** [comment]
> If anything, I would assume this to primarily be a latency issue, wherein regenerating a mapblock might take longer than a frame.

If not already done crack updates should jump ahead in the meshgen queue.

**Comment by Andrey2470T** [comment]
> Yes, this can definitely be optimized, but I wonder how important that is. The example in your post is deliberately contrived (and would arguably probably benefit most from some kind of instancing rather than batching). How bad is this phenomenon really for "normal" mapblocks?
> 
What do you mean "deliberately contrived"? I showed the edge case in my post where the bad current implementation of the crack animation makes the distinctly visible delays in that and negative affecting FPS.

> Under the assumption that mapblocks have similar average complexities, this single mapblock doesn't matter terribly much in the bigger picture when it comes to throughput. If we can't afford to regenerate a single mapblock mesh, how could we hope to generate the cubically many mapblock meshes around the player in reasonable time? If anything, I would assume this to primarily be a latency issue, wherein regenerating a mapblock might take longer than a frame.
>
Obviously it matters from my video for the reason that I described a bit above. And you don't need to regenerate many mapblocks meshes (that's a whole point of my suggestion).

**Comment by appgurueu** [comment]
> What do you mean "deliberately contrived"?

It's an artificial / unrealistic worst case where specifically the mapblock mesh which needs to be updated with the crack is a worst case for the mesher. From this it is not clear at all how much of an issue this is in practice, and there is reason to believe that it might not be much of an issue in practice, because generating the single mapblock mesh with the crack can be expected to be dominated by the generation of all the many other mapblock meshes.
(In other words: If generating the average mapblock mesh took 0.3s as in your contrived example, we'd have much bigger problems than the crack animation not updating in a timely manner.)

> Obviously it matters from my video for the reason that I described a bit above.

It matters in a contrived extreme scenario - that doesn't mean that it matters in practice, especially as your measurements e.g. with "average" MTG for example would indicate the opposite.
Hence why I'm asking you to what extent you experience this in "realistic" scenarios.

I understood the point of your suggestion, but the contrived example doesn't justify it.

> If not already done crack updates should jump ahead in the meshgen queue.

It looks like this is already the case (see the `urgent` bool in `addUpdateMeshTask`), though other mapblock mesh updates may use this bool as well; perhaps prioritization among them might be useful, but I haven't yet deeply looked into this.

**Comment by Andrey2470T** [comment]
> > What do you mean "deliberately contrived"?
> 
> It's an artificial / unrealistic worst case where specifically the mapblock mesh which needs to be updated with the crack is a worst case for the mesher. From this it is not clear at all how much of an issue this is in practice, and there is reason to believe that it might not be much of an issue in practice, because generating the single mapblock mesh with the crack can be expected to be dominated by the generation of all the many other mapblock meshes. (In other words: If generating the average mapblock mesh took 0.3s as in your contrived example, we'd have much bigger problems than the crack animation not updating in a timely manner.)
> 
It can be quite realistic and be in practice actually. My case is easy to obtain if you use special tools e.g. worldedit and generate that area with some complex meshes like on the video. Worldedit is known to be one of the top mods on CDB and on the forums, so the possibility of obtaining this scenario is in any case pretty not the low. Also, it re-generates not the only mesh, but still a few adjacent ones as it is seen from this code:
https://github.com/minetest/minetest/blob/f2ab887644f627ced3d91f589f7617cc0d647c27/src/client/client.cpp#L1746-L1765

Except there are two other benefits which I pointed in the first post: 1. avoiding the necessity to generate the combinations of the textures with the crack ones for each frame. It gives an opportunity not to overflow the texture source cache. 2. allowing the mix the cracks pixels with the underlying textures in the "overlay" blend mode.

**Comment by appgurueu** [comment]
> It can be quite realistic and be in practice actually. My case is easy to obtain if you use special tools e.g. worldedit and generate that area with some complex meshes like on the video. Worldedit is known to be one of the top mods on CDB and on the forums, so the possibility of obtaining this scenario is in any case pretty not the low.

This is still not a very practical statement. Yes, if people do very silly things, performance suffers. Worldedit users can be expected not to do that. It is obvious that filling a lot of space with complex meshes is likely to make performance suffer, and there is usually little motivation to do that (since inner nodes are mostly occluded, though the engine can't tell). And even then the crack is probably one of the "lesser" worries: Everything that modifies the mapblock will be delayed by the mapblock mesh regen time. Place a node? Change a param2? Have a mod swap a node (think mesecons blinky plant)? All of those won't work as they should, hence why I think the crack is only one concern of many.

The only way(s) to proper address this I see are roughly as follows: (1) don't, modders need to accept certain performance limitations (which in this case so far aren't unreasonable at all); (2) optimize mapblock meshgen to bump those limits; (3) perhaps let mapblock meshing subdivide expensive meshes further to speed up regen (at the cost of potentially issuing more drawcalls). Instancing might also be interesting to consider.

In other words: I don't think a change to cracked node handling specifically is justified for performance reasons here. It's a narrow patch for an extreme case problem from what I've seen so far.

I'd still be interested to know whether this is actually a real-world problem, e.g. have you played on a server (maybe modulo anarchy incl. cheats servers) where this was a problem, or a singleplayer game (as a player, not as a tester / developer), or do you know someone who did?

> Also, it re-generates not the only mesh, but still a few adjacent ones as it is seen from this code: [...]

Fair point, though doesn't change terribly much in the above analysis; it just adds a constant factor (which normally will be small, since the chance of being on a mapblock border isn't terribly high), and should easily be reducible when it comes to latency due to multithreading.

> 1. avoiding the necessity to generate the combinations of the textures with the crack ones for each frame. It gives an opportunity not to overflow the texture source cache.

Something like this seems like a potentially good motivation, and that's how I would spin it: As a dependency for other, more important performance improvements (which then need to be well motivated), and possibly as an opportunity to simplify the code?

> 2. allowing the mix the cracks pixels with the underlying textures in the "overlay" blend mode.

I'm not quite sure what the point of this is, since whatever overlay blending you can do on the GPU, you can also do on the CPU? (Besides, I'd prefer it if we didn't just change how cracks look for everyone.)

I'd also be worried that this causes incorrect crack rendering, because things are applied in the wrong order: First you render the node. There you already apply lighting etc. And only after that do you smash the crack on top. I doubt that this is in general equivalent to first putting the crack on the texture, then applying all the effects the fragment shader applies. It especially won't be if we are to e.g. readd bumpmapping in the future.

**Comment by Desour** [comment]
FYI, I've recently made something to track node positions of vertices, and turn mapblockmeshes back into prebuffers:
https://github.com/Desour/minetest/tree/tmp_meshgen_nodeidx
What's missing is passing a bitmap for the changed nodes and the old mesh to the meshgen thread, IIRC.
(Leaving this here, because I'm not planning on finishing it in the near future.)

**Comment by appgurueu** [comment]
A general way to efficiently incrementally update mapblock meshes makes sense (maybe open a separate issue to discuss that). It would be a neat side effect if that would make updating the cracked node easy.
